pandas/_libs/tslibs/timedeltas.pyx

```diff
@@ -229,10 +229,10 @@ cpdef inline int64_t cast_from_unit(object ts, object unit) except? -1:
     # cast the unit, multiply base/frace separately
     # to avoid precision issues from float -> int
     base = <int64_t> ts
-    frac = ts -base
+    frac = ts - base
     if p:
         frac = round(frac, p)
```

Comment:
you might be able to do something like this

``round(frac * 1000, p - 3) // 1000`` (for p >- 3) but haven't looked really closely.